### ISSUE TITLE
Cannot build index on windows

### ISSUE THREAD

**Issue by n00dl3** [buehler/typescript-hero#110]
#### System information
- OS: Windows
- TypeScript Hero version: 0.9
- VSCode version: 1.6.1
#### Description

Got an error while building the index :

> ```
> ResolveIndex: ERROR  Catched an error during buildIndex()
>     Data:   RangeError: Maximum call stack size exceeded
>     at RegExp.[Symbol.replace] (native)
>     at RegExp.[Symbol.replace] (native)
>     at String.replace (native)
>     at getNodeLibraryName (C:\Users\user\.vscode\extensions\rbbit.typescript-hero-0.9.0\out\src\caches\ResolveIndex.js:33:48)
>     at ResolveIndex.processResourceExports (C:\Users\user\.vscode\extensions\rbbit.typescript-hero-0.9.0\out\src\caches\ResolveIndex.js:199:33)
>     at ResolveIndex.processAllFromExport (C:\Users\user\.vscode\extensions\rbbit.typescript-hero-0.9.0\out\src\caches\ResolveIndex.js:225:14)
>     at ResolveIndex.processResourceExports (C:\Users\user\.vscode\extensions\rbbit.typescript-hero-0.9.0\out\src\caches\ResolveIndex.js:208:26)
>     at ResolveIndex.processAllFromExport (C:\Users\user\.vscode\extensions\rbbit.typescript-hero-0.9.0\out\src\caches\ResolveIndex.js:225:14)
>     at ResolveIndex.processResourceExports (C:\Users\user\.vscode\extensions\rbbit.typescript-hero-0.9.0\out\src\caches\ResolveIndex.js:208:26)
>     at ResolveIndex.processAllFromExport (C:\Users\user\.vscode\extensions\rbbit.typescript-hero-0.9.0\out\src\caches\ResolveIndex.js:225:14)
> ```
##### Steps to reproduce

Just run the "Rebuild index" command

**Comment by buehler** [comment]
Oh well. Seems like some library has like thousands of exports :-( or cyclic ones.
What frameworks do you use in your solution?

**Comment by n00dl3** [comment]
Angular 2 and a shitload of node_modules, i'll try to dig that when i get back to home, if I can because there's no such problem on linux and I encountered this on a colleague's computer.

----- Message de réponse -----
De : "Christoph Bühler" <notifications@github.com>
Pour : "buehler/typescript-hero" <typescript-hero@noreply.github.com>
Cc : "n00dl3" <julien.durens@yahoo.fr>, "Author" <author@noreply.github.com>
Objet : [buehler/typescript-hero] Cannot build index on windows (#110)
Date : lun., oct. 24, 2016 21:59

Oh well. Seems like some library has like thousands of exports :-( or cyclic ones.

What frameworks do you use in your solution?

&mdash;You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.










{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/buehler/typescript-hero","title":"buehler/typescript-hero","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/buehler/typescript-hero"}},"updates":{"snippets":[{"icon":"PERSON","message":"@buehler in #110: Oh well. Seems like some library has like thousands of exports :-( or cyclic ones.\r\nWhat frameworks do you use in your solution?\r\n"}],"action":{"name":"View Issue","url":"https://github.com/buehler/typescript-hero/issues/110#issuecomment-255848574"}}}

**Comment by quodroc** [comment]
I think the main problem is cycles. I modified `processResourceExports` to use a queue instead of recursion, which prevented the "maximum call stack" error but the resolver ran forever without finishing.

Next I tried keeping a list of resources that have already been processed which seems to fix the problem. Added the following to the top of `processResourceExports`
```
this.processed = this.processed || [];
if (this.processed.indexOf(resource) >= 0) {
	return;
}
this.processed.push(resource);
```
(code is hacky so I haven't done a proper pull request)
